mingpt/mingptmlx.py

```diff
@@ -0,0 +1,309 @@
+block_size = 32 # what is the maximum context length for predictions?
+max_iters = 2000
+eval_interval = 100
+learning_rate = 1e-3
+eval_iters = 200
+n_embd = 64
+n_head = 4
+n_layer = 4
+dropout = 0.01
+
+
+# %%
+# device = 'cuda' if torch.cuda.is_available() else 'cpu'
+device = mx.default_device()
+
+
+# %%
+# ------------
+
+mx.random.seed(1337)
+
+# wget https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt
+with open('input.txt', 'r', encoding='utf-8') as f:
+    text = f.read()
+
+
+# %%
+
+# here are all the unique characters that occur in this text
+chars = sorted(list(set(text)))
+vocab_size = len(chars)
+# create a mapping from characters to integers
+stoi = { ch:i for i,ch in enumerate(chars) }
+itos = { i:ch for i,ch in enumerate(chars) }
+encode = lambda s: [stoi[c] for c in s] # encoder: take a string, output a list of integers
+decode = lambda l: ''.join([itos[i] for i in l]) # decoder: take a list of integers, output a string
+
+
+# %%
+
+# Train and test splits
+data = mx.array(encode(text), dtype=mx.int64)
+n = int(0.9*len(data)) # first 90% will be train, rest val
+train_data = data[:n]
+val_data = data[n:]
+
+
+# %%
+
+# data loading
+def get_batch(split):
+    # generate a small batch of data of inputs x and targets y
+    data = train_data if split == 'train' else val_data
+    ix = mx.random.randint(0, len(data) - block_size, (batch_size,))
+    ix = [i.item() for i in ix]
+    x = mx.stack([data[i:i+block_size] for i in ix])
+    y = mx.stack([data[i+1:i+block_size+1] for i in ix])
+    # x, y = x.to(device), y.to(device)
+    return x, y
+
+
+# %%
+
+class Head(nn.Module):
+    """ one head of self-attention """
+
+    def __init__(self, head_size):
+        super().__init__()
+        self.key = nn.Linear(n_embd, head_size, bias=False)
+        self.query = nn.Linear(n_embd, head_size, bias=False)
+        self.value = nn.Linear(n_embd, head_size, bias=False)
+        self.tril = mx.tril(mx.ones([block_size, block_size]))
+        self.dropout = nn.Dropout(dropout)
+
+    def __call__(self, x):
+        B,T,C = x.shape
+        k = self.key(x)   # (B,T,C)
+        q = self.query(x) # (B,T,C)
+        # compute attention scores ("affinities")
+        wei = q @ k.transpose((0,2,1)) * C**-0.5 # (B, T, C) @ (B, C, T) -> (B, T, T)
+        mask = self.tril[:T, :T] == 0
+        wei = mx.where(mask, float('-inf'), wei) # (B, T, T)
+        wei = nn.softmax(wei, axis=-1) # (B, T, T)
+        wei = self.dropout(wei)
+        # perform the weighted aggregation of the values
+        v = self.value(x) # (B,T,C)
+        out = wei @ v # (B, T, T) @ (B, T, C) -> (B, T, C)
+        return out
+
+
+# %%
+
+class MultiHeadAttention(nn.Module):
+    """ multiple heads of self-attention in parallel """
+
+    def __init__(self, num_heads, head_size):
+        super().__init__()
+        self.heads = [Head(head_size) for _ in range(num_heads)]
```

Comment:
I would not do the heads this way. It should be way faster to do them in a single vectorized op.